src/Interpreters/ReplaceDistributedTableNameVisitor.cpp

```diff
@@ -0,0 +1,247 @@
+                    }
+                    else if (table_expr->table_function)
+                    {
+                        auto local_table_ident = enter(*table_expr->table_function->as<ASTFunction>(), table_expr->table_function, scope);
+                        if (local_table_ident)
+                        {
+                            table_expr->database_and_table_name = local_table_ident;
+                            table_expr->table_function.reset();
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+void ReplaceDistributedTableNameVisitor::leave(ASTSelectQuery &, ScopePtr & scope)
+{
+    /// Finish select query, reset scope.
+    scope = scope->parent_scope;
+}
+
+std::shared_ptr<ASTTableIdentifier>
+ReplaceDistributedTableNameVisitor::enter(ASTFunction & table_function, ASTPtr & table_function_ast, ScopePtr & /*scope*/)
+{
+    has_table_function = true;
+
+    if (table_function.name != "remote" || table_function.name != "remoteSecure" || table_function.name != "cluster" || table_function.name != "clusterAllReplicas") /// TODO cluster
+        return {};
+
+    TableFunctionPtr table_function_ptr = TableFunctionFactory::instance().get(table_function_ast, context);
+    StoragePtr storage = table_function_ptr->execute(table_function_ast, context, table_function_ptr->getName());
+
+    auto * storage_distributed = typeid_cast<StorageDistributed *>(storage.get());
+
+    auto database_name = storage_distributed->getRemoteDatabaseName();
+    auto table_name = storage_distributed->getRemoteTableName();
+    auto local_table_ident = std::make_shared<ASTTableIdentifier>(database_name, table_name);
+    auto local_table = local_table_ident->getTableId();
+
+    storages.emplace_back(local_table);
+    clusters.emplace_back(storage_distributed->getCluster());
+
+    if (auto sharding_key = storage_distributed->getShardingKey())
+    {
+        WriteBufferFromOwnString write_buffer;
+        IAST::FormatSettings settings(write_buffer, true, false, true);
+        sharding_key->format(settings);
+        sharding_keys.emplace_back(write_buffer.str());
+    }
+    else
+    {
+        sharding_keys.emplace_back();
+    }
+
+    has_distributed_table = true;
+
+    return local_table_ident;
+}
+
+void ReplaceDistributedTableNameVisitor::enter(ASTTableIdentifier & table_ident, ScopePtr & scope)
+{
+    if (!table_ident.getDatabase())
+        table_ident.resetTable(context->getCurrentDatabase(), table_ident.shortName());
+
+    StoragePtr table = DatabaseCatalog::instance().getTable(table_ident.getTableId(), context);
+
+    if (auto * distributed_table = dynamic_cast<StorageDistributed *>(table.get()))
+    {
+        /// 1. Initialize scope
+        if (std::find(scope->tables.begin(), scope->tables.end(), table_ident.getTableId()) != scope->tables.end())
+            throw Exception(ErrorCodes::SYNTAX_ERROR, "duplicated table {}", table_ident.shortName());
```

Comment:
this fails for a self join, lets say i have a distributed table table1 that i want to join with itself on some column conditions, this fails.

eg. query: select count(*) from table1 as t1 join table1 as t2 on t1.c1 = t2.c1 * 2

Reply:
Thanks for your review and report.  Fixed